src/libponyrt/sched/systematic_testing.c

```diff
@@ -0,0 +1,217 @@
+#define PONY_WANT_ATOMIC_DEFS
+
+#include "systematic_testing.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include "cpu.h"
+#include "ponyassert.h"
+
+typedef struct systematic_testing_thread_t
+{
+  pony_thread_id_t tid;
+  pony_signal_event_t sleep_object;
+  bool stopped;
+} systematic_testing_thread_t;
+
+static systematic_testing_thread_t* active_thread = NULL;
+static systematic_testing_thread_t* threads_to_track = NULL;
+static uint32_t total_threads = 0;
+static uint32_t stopped_threads = 0;
+static PONY_ATOMIC(uint32_t) waiting_to_start_count;
+
+#if defined(USE_SCHEDULER_SCALING_PTHREADS)
+static pthread_mutex_t systematic_testing_mut;
+
+static pthread_once_t systematic_testing_mut_once = PTHREAD_ONCE_INIT;
+
+void systematic_testing_mut_init()
+{
+  pthread_mutex_init(&systematic_testing_mut, NULL);
+}
+#endif
+
+void ponyint_systematic_testing_init(uint64_t random_seed, uint32_t max_threads)
+{
+  #if defined(USE_SCHEDULER_SCALING_PTHREADS)
+  pthread_once(&systematic_testing_mut_once, systematic_testing_mut_init);
+#endif
+
+  atomic_store_explicit(&waiting_to_start_count, 0, memory_order_relaxed);
+
+  if(0 == random_seed)
+    random_seed = ponyint_cpu_tick();
+
+  SYSTEMATIC_TESTING_PRINTF("Systematic testing using seed: %lu...\n", random_seed);
+  SYSTEMATIC_TESTING_PRINTF("(rerun with `<app> --ponysystematictestingseed %lu` to reproduce)\n", random_seed);
+
+  // TODO: maybe replace with better RNG?
+  // probably best to replace this with a RNG that has the exact same behavior
+  // on all platform so we can deterministically reproduce the same thread
+  // interleaving everywhere if needed
+  srand((int)random_seed);
+
+  // initialize thead tracking array (should be max_threads + 1 to account for asio)
+  // TODO: memtrack tracking for these allocations
```

Comment:
Can you include this enhancement in the current PR?

Reply:
done